File: android/app/src/main/java/com/david/tehilim/core/persistence/ChainArchiveStore.kt
package com.david.tehilim.core.persistence

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val Context.chainsDataStore by preferencesDataStore(name = "tehilim_chains")

/** Snapshot local d'une chaîne (compte rendu), conservé par le créateur. */
@Serializable
data class ChainArchiveSnapshot(
    val id: String,
    val name: String,
    val intentionWire: String,
    val detail: String,
    val creatorName: String,
    val readingDeadlineMillis: Long,
    val archivedAtMillis: Long,
    /** psalmId ("1".."150") → nom du lecteur. */
    val assignments: Map<String, String>
) {
    val subjectLine: String
        get() = if (detail.isBlank()) name else "$name — $detail"
}

/**
 * Persiste localement : (1) les ids de chaînes connues (créées/rejointes) pour
 * « Mes chaînes » ; (2) les archives créateur. Mirror Android de
 * ChainArchiveStore.swift. DataStore + JSON (kotlinx.serialization).
 */
class ChainArchiveStore(private val context: Context) {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val knownKey = stringPreferencesKey("chains.known.ids")
    private val archivesKey = stringPreferencesKey("chains.archives")
    private val json = Json { ignoreUnknownKeys = true }

    private val _knownChainIds = MutableStateFlow<List<String>>(emptyList())
    val knownChainIds: StateFlow<List<String>> = _knownChainIds

    private val _archives = MutableStateFlow<List<ChainArchiveSnapshot>>(emptyList())
    val archives: StateFlow<List<ChainArchiveSnapshot>> = _archives

    init {
        scope.launch { loadInitial() }
    }

    private suspend fun loadInitial() {
        val prefs = context.chainsDataStore.data.first()
        prefs[knownKey]?.let { raw ->
            runCatching { json.decodeFromString<List<String>>(raw) }.getOrNull()?.let {
                _knownChainIds.value = it
            }
        }
        prefs[archivesKey]?.let { raw ->
            runCatching { json.decodeFromString<List<ChainArchiveSnapshot>>(raw) }.getOrNull()?.let {
                _archives.value = it
            }
        }
    }

    fun remember(chainId: String) {
        if (chainId.isBlank()) return
        val next = (listOf(chainId) + _knownChainIds.value.filter { it != chainId })
        _knownChainIds.value = next
        scope.launch { persistKnown(next) }
    }

    fun forget(chainId: String) {
        val next = _knownChainIds.value.filter { it != chainId }
        _knownChainIds.value = next
        scope.launch { persistKnown(next) }
    }

    fun saveArchive(snapshot: ChainArchiveSnapshot) {
        val next = listOf(snapshot) + _archives.value.filter { it.id != snapshot.id }
        _archives.value = next
        scope.launch { persistArchives(next) }
    }

    fun deleteArchive(id: String) {
        val next = _archives.value.filter { it.id != id }
        _archives.value = next
        scope.launch { persistArchives(next) }
    }

    private suspend fun persistKnown(ids: List<String>) {
        context.chainsDataStore.edit { it[knownKey] = json.encodeToString(ids) }
    }

    private suspend fun persistArchives(list: List<ChainArchiveSnapshot>) {
        context.chainsDataStore.edit { it[archivesKey] = json.encodeToString(list) }
    }
}

File: ios/Tehilim/Core/Persistence/ChainArchiveStore.swift
import Foundation

/// Snapshot local d'une chaîne (compte rendu final), conservé par le **créateur**
/// même après la suppression cloud (TTL). Permet de garder l'historique.
struct ChainArchiveSnapshot: Codable, Identifiable, Equatable {
    let id: String
    let name: String
    let intentionRaw: String
    let detail: String
    let creatorName: String
    let readingDeadline: Date
    let archivedAt: Date
    /// psalmId ("1".."150") → nom du lecteur.
    let assignments: [String: String]

    var intention: ChainIntention { ChainIntention(rawValue: intentionRaw) ?? .reussite }
    var subjectLine: String {
        let d = detail.trimmingCharacters(in: .whitespacesAndNewlines)
        return d.isEmpty ? name : "\(name) — \(d)"
    }
}

/// Persiste localement : (1) les ids de chaînes connues (créées/rejointes) pour
/// alimenter « Mes chaînes » sans requête globale ; (2) les archives créateur.
final class ChainArchiveStore: ObservableObject {
    @Published private(set) var knownChainIds: [String] = []
    @Published private(set) var archives: [ChainArchiveSnapshot] = []

    private let defaults: UserDefaults
    private let knownKey = "chains.known.ids"
    private let archivesKey = "chains.archives"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        if let ids = defaults.array(forKey: knownKey) as? [String] {
            knownChainIds = ids
        }
        if let data = defaults.data(forKey: archivesKey),
           let decoded = try? JSONDecoder().decode([ChainArchiveSnapshot].self, from: data) {
            archives = decoded
        }
    }

    /// Mémorise une chaîne (créée ou rejointe) — en tête, sans doublon.
    func remember(_ chainId: String) {
        guard !chainId.isEmpty else { return }
        knownChainIds.removeAll { $0 == chainId }
        knownChainIds.insert(chainId, at: 0)
        defaults.set(knownChainIds, forKey: knownKey)
    }

    func forget(_ chainId: String) {
        knownChainIds.removeAll { $0 == chainId }
        defaults.set(knownChainIds, forKey: knownKey)
    }

    /// Ajoute (ou remplace) une archive de compte rendu.
    func saveArchive(_ snapshot: ChainArchiveSnapshot) {
        archives.removeAll { $0.id == snapshot.id }
        archives.insert(snapshot, at: 0)
        persistArchives()
    }

    func deleteArchive(_ id: String) {
        archives.removeAll { $0.id == id }
        persistArchives()
    }

    private func persistArchives() {
        if let data = try? JSONEncoder().encode(archives) {
            defaults.set(data, forKey: archivesKey)
        }
    }
}
